src/rebar_app_discover.erl

```diff
@@ -171,17 +171,35 @@ project_app_config(AppInfo, State) ->
 maybe_reset_hooks(Dir, Opts, State) ->
     case ec_file:real_dir_path(rebar_dir:root_dir(State)) of
         Dir ->
-            reset_hooks(Opts);
+            CurrentProfiles = rebar_state:current_profiles(State),
+            reset_hooks(Opts, CurrentProfiles);
         _ ->
             Opts
     end.
 
 %% @doc make the hooks empty for a given set of options
--spec reset_hooks(Opts) -> Opts when Opts :: rebar_dict().
-reset_hooks(Opts) ->
-    lists:foldl(fun(Key, OptsAcc) ->
-                        rebar_opts:set(OptsAcc, Key, [])
-                end, Opts, [post_hooks, pre_hooks, provider_hooks, artifacts]).
+-spec reset_hooks(Opts, Profiles) ->
+    Opts when
+      Opts :: rebar_dict(),
+      Profiles :: [atom()].
+reset_hooks(Opts, CurrentProfiles) ->
+    Opts1 = lists:foldl(fun(Key, OptsAcc) ->
+                            rebar_opts:set(OptsAcc, Key, [])
+                        end, Opts, [post_hooks, pre_hooks, provider_hooks, artifacts]),
+    Profiles = rebar_opts:get(Opts1, profiles, []),
+    Profiles1 = lists:map(fun({P, ProfileOpts}) ->
+                              case lists:member(P, CurrentProfiles) of
+                                  true ->
+                                      {P, [X || X={Key, _} <- ProfileOpts,
+                                                not lists:member(Key, [post_hooks,
+                                                                       pre_hooks,
+                                                                       provider_hooks,
+                                                                       artifacts])]};
```

Comment:
I'd extract that list of magic keys to escape here and on line 188 to some shared variable or macro so they always end up matching.